### PR TITLE
[experimental] package-management repository configuration

### PR DESCRIPTION
**Pull request by cjp256**
repo: add installation support for package-repositories

- Enable installation support for PPAs.

  - Lookup GPG key fingerprint from launchpad using launchpadlib.

  - Install source using install_sources() so it is treated the
    same as other requested sources, but is a nice shortcut for
    the user.

- Enable installation of GPG keys from local or keyserver sources.

- Allow user to specify `key-id` which can either be the GPG key
  fingerprint/ID, or an identifier matching a key in the assets dir.

- Smarter refresh apt cache logic, only refreshing when when either
  the apt configuration or GPG keys are modified after a batch install.
  That is, all user-configured repositories are installed, then apt
  will refresh the cache.  This batch is currently separate from the
  process used by the ROS plugins to install their repositories. Future
  will make this more consistent.

- Remove now-unused _get_local_sources_list().

- Prefix default names for apt sources with 'snapcraft-'.

- Add unit test coverage and three spread tests to cover:

  1. installation of PPA

  2. installation of source using keyserver

  3. installation of source using key assets via fingerprint

  4. installation of source using key assets via name

### CODE REVIEW

**Comment by casperdcl** [comment]
seems to work fine here https://github.com/casperdcl/cli/blob/snap/snap/snapcraft.yaml -> https://snapcraft.io/gh

**Comment by casperdcl** [comment]
ah there's an issue with subsequent runs:

```
$ snap refresh snapcraft --channel=edge/pr-2911
...
$ snapcraft
... # all works fine
$ snapcraft  # second run
Launching a VM.
snapd 2.43.2 from Canonical✓ refreshed
Issues while validating snapcraft.yaml: Additional properties are not allowed ('package-management' was unexpected)
Run the same command again with --debug to shell into the environment if you wish to introspect this failure.
$ snapcraft clean
$ snapcraft # works again

**Comment by cjp256** [comment]
@casperdcl Thank you for the report.  I cannot seem to reproduce your specific error.  Can you share your YAML?  Is it still reproducible for you?

**Comment by casperdcl** [comment]
Odd; can't reproduce anymore.

**Comment by cjp256** [comment]
> Odd; can't reproduce anymore.

@casperdcl: FYI - there has been a bug reported which explains the behavior you saw previously. https://bugs.launchpad.net/snapcraft/+bug/1865834  We expect to fix it soon.

**Comment by cjp256** [comment]
rebased to addressed conflicts

**Comment by cjp256** [comment]
Now requires https://github.com/snapcore/snapcraft/pull/2970 to enable experimental flag

**Comment by sergiusens** [comment]
Seems like legit broken cases in

    - google:ubuntu-16.04-64:tests/spread/plugins/v1/catkin/catkin-with-python

    - google:ubuntu-16.04-64:tests/spread/plugins/v1/catkin/content-sharing

    - google:ubuntu-16.04-64:tests/spread/plugins/v1/catkin/pip-support

    - google:ubuntu-16.04-64:tests/spread/plugins/v1/catkin/recursive-rosinstall

    - google:ubuntu-16.04-64:tests/spread/plugins/v1/catkin/run

    - google:ubuntu-16.04-64:tests/spread/plugins/v1/catkin/source-dependencies

    - google:ubuntu-18.04-64:tests/spread/plugins/v1/catkin/catkin-with-python

    - google:ubuntu-18.04-64:tests/spread/plugins/v1/catkin/content-sharing

    - google:ubuntu-18.04-64:tests/spread/plugins/v1/catkin/pip-support

    - google:ubuntu-18.04-64:tests/spread/plugins/v1/catkin/recursive-rosinstall

    - google:ubuntu-18.04-64:tests/spread/plugins/v1/catkin/run

    - google:ubuntu-18.04-64:tests/spread/plugins/v1/catkin/source-dependencies

    - google:ubuntu-18.04-64:tests/spread/plugins/v1/catkin/stage-snaps

    - google:ubuntu-18.04-64:tests/spread/plugins/v1/colcon/run

    - google:ubuntu-20.04-64:tests/spread/general/package-repositories:test_apt_deb

    - google:ubuntu-20.04-64:tests/spread/general/package-repositories:test_apt_keyserver

    - google:ubuntu-20.04-64:tests/spread/general/package-repositories:test_apt_ppa

Some refer to commands without a relative path within the prime directory.

**Comment by cjp256** [comment]
A couple TODOs:
1) spread restore should revert to previous apt + gpg configuration
2) gnupg library is logging to info that we need to squelch

File: snapcraft/internal/repo/_deb.py [hunk 1]
```diff
@@ -558,10 +604,98 @@ def install_gpg_key(cls, *, key_id: str, key: str) -> None:
...
+            "--keyserver",
+            key_server,
+            "--recv-keys",
+            key_id,
+        ]
+
+        try:
+            logger.debug(f"Executing: {cmd!r}")
+            subprocess.run(
+                cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, check=True
```
Comment (sergiusens) [inline snapcraft/internal/repo/_deb.py:643]: pass in env with LANG C.UTF-8

Comment (cjp256) [reply snapcraft/internal/repo/_deb.py:646]: done

**Comment by sergiusens** [comment]
Store errors again.

**Pull request merged**